### DISCUSSION TITLE
Rockwren 1.0.0 released: a micropython device framework that simplifies the creation of connected devices

### DISCUSSION THREAD

**Discussion by ccrighton** [micropython/micropython#12641] [Show and tell]
### Micropython Rockwren 1.0.0

Use [Rockwren](https://github.com/ccrighton/rockwren/blob/main/README.md) to connect your Pico W or similar device via web and mqtt interfaces with two lines of code in ```main.py```:

```
from rockwren import rockwren

rockwren.fly(rockwren.Device())
```

<img width="200" alt="main-screen" src="https://github.com/micropython/micropython/assets/6871114/4214e48b-64bf-4ae2-b9b7-e740a7d92b65">
<img width="193" alt="picow_temperature_example" src="https://github.com/micropython/micropython/assets/6871114/17bafabd-957d-476b-8e4f-2bfcc86edc8b">

### Features

- [APIs](https://github.com/ccrighton/rockwren/blob/main/docs/apis.md) for device creation
- Web interface for configuration and device control
- MQTT client with certificate support (e.g. AWS IoT or [Mosquitto](https://mosquitto.org/))
- Access Point mode for device configuration
- Home Assistant discovery via MQTT
- The web interface supports:
  - An extendable web interface for device specific controls and updating status display
  - MQTT, WiFi configuration
  - A log viewer
  - A device information viewer

### Install

Install the [package](https://pypi.org/project/micropython-rockwren/) ```micropython-rockwren``` with Thonny, pipkin or another package management tool.

### Home Assistant
[Home Assistant integration](https://github.com/ccrighton/rockwren/blob/main/docs/home-assistant-discovery.md) with discovery made easy:

<img width="523" alt="home-assistant-picow-led-device" src="https://github.com/micropython/micropython/assets/6871114/e6eaad04-8648-4944-8911-dc0b2a08de01">

### Tutorial

Check out the [tutorial](https://github.com/ccrighton/rockwren/blob/main/docs/picow-led-light-tutorial.md) for creating a Pico W LED light integrated with Home Assistant and with a web interface.

### Examples

- [Pico W LED Light](https://github.com/ccrighton/rockwren/blob/main/examples/pico_led_light/main.py)
- [Pico W Binary Sensor](https://github.com/ccrighton/rockwren/blob/main/examples/pico_binary_sensor/main.py)
- [Pico W Temperature Sensor](https://github.com/ccrighton/rockwren/blob/main/examples/pico_temperature/main.py)
- [Pico W Switch](https://github.com/ccrighton/rockwren/blob/main/examples/pico_switch/main.py)

### Documentation

https://github.com/ccrighton/rockwren/blob/main/docs/TOC.md

**Comment by andrewleech** [comment]
This looks really neat, thanks for sharing! I've been wanting an easy way to use micropython for quick HA sensors, lights, controls etc for some time. 
I've used [pysmartnode](https://github.com/kevinkk525/pysmartnode) a couple of times in the past but it's _big_ with a lot of layers, quite hard to build something custom and small with it.

I've also played with some captive-portal stuff before for use in "Access Point mode for device configuration" which I think is _really_ useful for these applications, I'll have a got at integrating it at some point when I next have an application I want to try building :-D

Was there anything in particular that pushed you towards full GPL for this project? It's pretty rare for embedded applications as it ends up excluding _many_ use cases due to the fact everything is linked into one application rather than a collection of separate applications like on linux. We _cannot_ use gpl embedded libraries at work for instance even as a company that contributes to open source projects frequently. 
Don't get me wrong, it's your project so completely your call, but in my experience it can be overly restrictive. Maybe it's not so bad for an application framework rather than a library though.

**Reply by ccrighton** [reply]
I'd be interested in getting your feedback when you have a chance to use it.

The access point config mode it pretty useful for the devices I've built using Rockwren: a RGBW LED controller, a garage door controller and a ventilation controller.  At the moment the implementation just allows the SSID and password to be set.  It could do more but that seems sufficient for now.

I've currently licensed it under the GPL.  I'd noticed some other frameworks like Rockwren also used the GPL so thought it would be a good starting point.  I have the copyright to all the code (unless marked otherwise) so I may consider re-licensing to a more permissive license at some point in the future.  I've licensed the documentation under CC-BY-4.0 to allow commercial remixing.  The modified [Pimoroni Phew](https://github.com/ccrighton/phew) web library is licensed under MIT.

**Reply by andrewleech** [reply]
Sounds good. Also I was curious about the use of `phew`, I thought everyone uses `microdot` already :-) ? Does `phew` provide anything extra / different that's useful, or was it just the one available when you started this project?

**Reply by ccrighton** [reply]
Phew is a bit problematic. It had some critical bugs hence the fork. It was what I was using so didn't look much further. I'll check out microdot. Thanks for the suggestion.
